File: IP/src/iProlog/Toks.java
package iProlog;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * Reads chars from char streams using the current default encoding
 */
public class Toks extends StreamTokenizer {

    // reserved words - with syntactic function

    private static final String IF = "if";
    private static final String AND = "and";
    private static final String DOT = ".";
    private static final String HOLDS = "holds";
    private static final String LISTS = "lists"; // todo
    private static final String IS = "is"; // todo

    private Toks(Reader reader) {
        super(reader);
        resetSyntax();
        eolIsSignificant(false);
        ordinaryChar('.');
        ordinaryChars('!', '/'); // 33-47
        ordinaryChars(':', '@'); // 55-64
        ordinaryChars('[', '`'); // 91-96
        ordinaryChars('{', '~'); // 123-126
        wordChars('_', '_');
        wordChars('a', 'z');
        wordChars('A', 'Z');
        wordChars('0', '9');
        slashStarComments(true);
        slashSlashComments(true);
        ordinaryChar('%');
    }

    private static Toks toks(String s, boolean fromFile) {
        try {
            Reader R;
            R = fromFile ? new FileReader(s) : new StringReader(s);
            Toks T = new Toks(R);
            return T;

        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static ArrayList<ArrayList<ArrayList<String>>> sentences(String s, boolean fromFile) {
        ArrayList<ArrayList<ArrayList<String>>> Wsss = new ArrayList<>();
        ArrayList<ArrayList<String>> Wss = new ArrayList<>();
        ArrayList<String> Ws = new ArrayList<>();
        Toks toks = toks(s, fromFile);
        String t = null;
        while (null != (t = toks.getWord())) {

            if (DOT.equals(t)) {
                Wss.add(Ws);
                Wsss.add(Wss);
                Wss = new ArrayList<>();
                Ws = new ArrayList<>();
            } else if (("c:" + IF).equals(t)) {

                Wss.add(Ws);

                Ws = new ArrayList<>();
            } else if (("c:" + AND).equals(t)) {
                Wss.add(Ws);

                Ws = new ArrayList<>();
            } else if (("c:" + HOLDS).equals(t)) {
                String w = Ws.get(0);
                Ws.set(0, "h:" + w.substring(2));
            } else if (("c:" + LISTS).equals(t)) {
                String w = Ws.get(0);
                Ws.set(0, "l:" + w.substring(2));
            } else if (("c:" + IS).equals(t)) {
                String w = Ws.get(0);
                Ws.set(0, "f:" + w.substring(2));
            } else {
                Ws.add(t);
            }
        }
        return Wsss;
    }

    static String toString(Object[] Wsss) {
        return Arrays.deepToString(Wsss);
    }

    public static void main(String[] args) {

        Main.pp(sentences("prog.nl", true));
    }

    private String getWord() {
        String t = null;

        int c = StreamTokenizer.TT_EOF;
        try {
            c = nextToken();
            while (Character.isWhitespace(c) && c != StreamTokenizer.TT_EOF) {
                c = nextToken();
            }
        } catch (IOException ignored) {
            return "*** tokenizer error:" + ignored;
        }

        switch (c) {
            case StreamTokenizer.TT_WORD: {
                char first = sval.charAt(0);
                if (Character.isUpperCase(first) || '_' == first) {
                    t = "v:" + sval;
                } else {
                    try {
                        int n = Integer.parseInt(sval);
                        t = Math.abs(n) < 1 << 28 ? "n:" + sval : "c:" + sval;
                    } catch (Exception ignored) {
                        t = "c:" + sval;
                    }
                }
            }
            break;

            case StreamTokenizer.TT_EOF: {
                t = null;
            }
            break;

            default: {
                t = "" + (char) c;
            }

        }
        return t;
    }
}

File: JS/src/Toks.js


// Should this extend something? What does StreamTokenizer do?
// var Toks = class Toks {

  IF = "if";
  AND = "and";
  DOT = ".";
  HOLDS = "holds";
  LISTS = "lists";
  IS = "is";

  /*
  * This function takes a string and reads it into a variable
  *
  * TODO: read how to take in text from a file in node.js
  *
  * @param s - a string
  * @param fromFile - boolean to determine if the string is from a file
  *
  * https://github.com/Floby/node-tokenizer
  *
  */
  function makeToks(s, fromFile){
	  
    // Read s from a file
    s = require('fs');
    s.readFile('/' + fromFile, 'utf8', function(err,data){
      if(err){
        return console.log(err);
        }
        else{
          console.log(data);
        }
    });
    
    var T = new Toks(s);
    return T;
  }

  /*
  * TODO: research the functions used in this method
  *
  * @param R - The read in string from makeToks
  */
  function Toks(s){
    var Tokenizer = require('tokenizer');
    var mycallback;
    var t = new Tokenizer(mycallback);

    t.addRule(/^.$/, 'ordinary');
    t.addRule(/^!$/, 'ordinary');
    t.addRule(/^"$/, 'ordinary');
    t.addRule(/^#$/, 'ordinary');
    t.addRule(/^\$$/, 'ordinary');
    t.addRule(/^%$/, 'ordinary');
    t.addRule(/^&$/, 'ordinary');
    t.addRule(/^'$/, 'ordinary');
    t.addRule(/^\($/, 'ordinary');
    t.addRule(/^\)$/, 'ordinary');
    t.addRule(/^\*$/, 'ordinary');
    t.addRule(/^\+$/, 'ordinary');
    t.addRule(/^,$/, 'ordinary');
    t.addRule(/^-$/, 'ordinary');
    t.addRule(/^\/$/, 'ordinary');
    t.addRule(/^:$/, 'ordinary');
    t.addRule(/^;$/, 'ordinary');
    t.addRule(/^<$/, 'ordinary');
    t.addRule(/^=$/, 'ordinary');
    t.addRule(/^>$/, 'ordinary');
    t.addRule(/^\?$/, 'ordinary');
    t.addRule(/^@$/, 'ordinary');
    t.addRule(/^\[$/, 'ordinary');
    t.addRule(/^\\$/, 'ordinary');
    t.addRule(/^]$/, 'ordinary');
    t.addRule(/^\^$/, 'ordinary');
    t.addRule(/^_$/, 'ordinary');
    t.addRule(/^`$/, 'ordinary');
    t.addRule(/^{$/, 'ordinary');
    t.addRule(/^\|$/, 'ordinary');
    t.addRule(/^}$/, 'ordinary');
    t.addRule(/^~$/, 'ordinary');
    t.addRule(/^%$/, 'ordinary');

    t.addRule(/^a$/, 'wordchars');
    t.addRule(/^b$/, 'wordchars');
    t.addRule(/^c$/, 'wordchars');
    t.addRule(/^d$/, 'wordchars');
    t.addRule(/^e$/, 'wordchars');
    t.addRule(/^f$/, 'wordchars');
    t.addRule(/^g$/, 'wordchars');
    t.addRule(/^h$/, 'wordchars');
    t.addRule(/^i$/, 'wordchars');
    t.addRule(/^j$/, 'wordchars');
    t.addRule(/^k$/, 'wordchars');
    t.addRule(/^l$/, 'wordchars');
    t.addRule(/^m$/, 'wordchars');
    t.addRule(/^n$/, 'wordchars');
    t.addRule(/^o$/, 'wordchars');
    t.addRule(/^p$/, 'wordchars');
    t.addRule(/^q$/, 'wordchars');
    t.addRule(/^r$/, 'wordchars');
    t.addRule(/^s$/, 'wordchars');
    t.addRule(/^t$/, 'wordchars');
    t.addRule(/^u$/, 'wordchars');
    t.addRule(/^v$/, 'wordchars');
    t.addRule(/^w$/, 'wordchars');
    t.addRule(/^x$/, 'wordchars');
    t.addRule(/^y$/, 'wordchars');
    t.addRule(/^z$/, 'wordchars');

    t.addRule(/^A$/, 'wordchars');
    t.addRule(/^B$/, 'wordchars');
    t.addRule(/^C$/, 'wordchars');
    t.addRule(/^D$/, 'wordchars');
    t.addRule(/^E$/, 'wordchars');
    t.addRule(/^F$/, 'wordchars');
    t.addRule(/^G$/, 'wordchars');
    t.addRule(/^H$/, 'wordchars');
    t.addRule(/^I$/, 'wordchars');
    t.addRule(/^J$/, 'wordchars');
    t.addRule(/^K$/, 'wordchars');
    t.addRule(/^L$/, 'wordchars');
    t.addRule(/^M$/, 'wordchars');
    t.addRule(/^N$/, 'wordchars');
    t.addRule(/^O$/, 'wordchars');
    t.addRule(/^P$/, 'wordchars');
    t.addRule(/^Q$/, 'wordchars');
    t.addRule(/^R$/, 'wordchars');
    t.addRule(/^S$/, 'wordchars');
    t.addRule(/^T$/, 'wordchars');
    t.addRule(/^U$/, 'wordchars');
    t.addRule(/^V$/, 'wordchars');
    t.addRule(/^W$/, 'wordchars');
    t.addRule(/^X$/, 'wordchars');
    t.addRule(/^Y$/, 'wordchars');
    t.addRule(/^Z$/, 'wordchars');
    
    t.addRule(/^0$/, 'wordchars');
    t.addRule(/^1$/, 'wordchars');
    t.addRule(/^2$/, 'wordchars');
    t.addRule(/^3$/, 'wordchars');
    t.addRule(/^4$/, 'wordchars');
    t.addRule(/^5$/, 'wordchars');
    t.addRule(/^6$/, 'wordchars');
    t.addRule(/^7$/, 'wordchars');
    t.addRule(/^8$/, 'wordchars');
    t.addRule(/^9$/, 'wordchars');
   
   
    t.addRule(/^\/\/$/, 'comment');
    t.addRule(/^\\*$/, 'comment');
    t.addRule(/^\*\/$/, 'comment');

    //slashStarComments(true)
    //slashSlashComments(true)
	  
  }

  function getWord(){
    var t = "";
    
    var eos = require('end-of-stream');

    try{
      c = nextToken;
      while(c == " " && c != TT_EOF){
        c = nextToken;
      }
    } 
    catch(err){
      return "*** tokenizer error:" + t;
    }

    switch(c)
    {
      case TT_WORD:
        var first = sval.charAt(0);
        if(first == first.toUpperCase() || '_' == first){
          t = "v:" + sval;
        } else {
          try{
            var n = parseInt(sval);
            if(Math.abs(n) < 1 << 28){
              t = "n:" + sval;
            }else{
              t = "c:" + sval;
            }
          }
          catch(err){
            t = "c:" + sval;
          }
        }
      break;
      case eos:
        t = null;
      break;
      default:
        t = "" + c;
    }
    return t;
  }

  function toSentences(s, fromFile){
    var Wsss = {};
    var Wss = {};
    var Ws = [];
    var w = "";

    toks = makeToks(s, fromFile);
    t = "";
    while(null != (t = getWord())){
      if(DOT == t){
        Wss.push(Ws);
        Wsss.push(Wss);
        Wss = {};
        ws = [];
      }
      else if((("c:") + IF) == t){
        Wss.add(Ws);
        Ws = [];
      }
      else if((("c:") + AND) == t){
        Wss.push(Ws);
        Ws = [];
      }
      else if((("c:") + HOLDS) == t){
        w = Ws[0];
        Ws.splice(0, 0, "h: " + w.substring(2));
      }
      else if((("c:") + LISTS) == t){
        w = Ws[0];
        Ws.splice(0,0, "l:" + w.substring(2));
      }
      else if((("c:") + IS) == t){
        w = Ws[0];
        Ws.splice(0,0,"f:" + w.substring(2));
      }
      else{
        Ws.push(t);
      }
    }
    return Wsss;
  }

  function toString(Wsss){
    console.log(Wsss);
  }

  exports.main = function(){
    return true;
    toSentences("prog.nl", true);
  }

